Adding process tracing

akoutmos opened the pull request:
This PR implements #159 

Continuing the discussion from that issue:

> Should we not show the bars if the process already exists or should we show the bar from beginning to end?

At the moment I have the bar starting when a process is spawned and stopped when it exits (both of which have to be present during the trace in order for the bar to appear). In other words, it doesn't look like Mermaid renders anything unless there is both an actor activate and deactivate. That and it doesn't look like we can start bars at arbitrary points on the sequence. Rather it looks like they have to start and stop on events. All this to say, *I think* if we want to keep the bars, they'll only be present when a process starts and stops during the trace.

> Regarding the messages, can we do any sort of mouse over thing? Keep only the label and when you mouse over, you see the whole message?

I'll have to explore this a bit more. I tried setting up a pop up [Actor Menu](https://mermaid-js.github.io/mermaid/#/sequenceDiagram?id=actor-menus) as described in the docs...but the menu never popped up lol.

> There are some dotted arrows after the messages... is this the "receive" event? If so, I think we can ignore it. Another option would be for the arrows to not be horizontal, but point to when the messages are received in the other process timeline, but that may be too complicated?

I think the Mermaid diagrams only allow for horizontal lines and cannot be linked back to another event. Do you want me to remove all receive events and just have send events? Or only in certain circumstances? Here is an example with no receives for reference:

![image](https://user-images.githubusercontent.com/4753634/176985051-3e258d7e-39f8-4569-b43f-d5e2822c795a.png)

>We also need to check if we trace :DOWN messages from Process.monitor.

It has been done! I also added that to the example code :).

I have a couple more things that I need to clean up and fix so I am opening this up as a draft PR. Should have those wrapped up during the weekend :). Thanks in advance for the review and have a great weekend!

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,205 @@ defmodule Kino.Process do
+    trace_function.()
+
+    # Reset all of the tracing options
+    :erlang.trace(:all, false, [:send, :receive])
```
I think we don't need to change Erlang trace, do we?

akoutmos replied:
Ooops. That was copy and paste carry over from when I was experimenting with trace versus seq_trace. Fixed!

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,205 @@ defmodule Kino.Process do
+
+    # Reset all of the tracing options
+    :erlang.trace(:all, false, [:send, :receive])
+    :seq_trace.set_system_tracer(previous_tracer)
+    :seq_trace.reset_trace()
```
We should likely wrap the function in a `try` and execute this in `after`. :)

josevalim replied:
Likely something like this:

```elixir
{res, trace_events} =
  try do
    {trace_function.(), Tracer.get_trace_events(tracer_pid)}
  after
    Tracer.stop(tracer_pid)
    :seq_trace.set_system_tracer(previous_tracer)
    :seq_trace.reset_trace()
  end
```

So you ensure no seq_trace nor process leak.

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,205 @@ defmodule Kino.Process do
+    case message do
+      {:EXIT, _, reason} -> "EXIT: #{label_from_reason(reason)}"
+      {:spawn_request, _, _, _, _, _, _, _} -> "SPAWN"
+      {:down, _, reason} -> "DOWN: #{label_from_reason(reason)}"
```
```suggestion
      {:DOWN, _, :process, _, reason} -> "DOWN: #{label_from_reason(reason)}"
```

The label that you were before was specific to Task.async_stream. :)

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,205 @@ defmodule Kino.Process do
+       when is_tuple(tuple) and is_atom(elem(tuple, 0)),
+       do: elem(tuple, 0)
+
+  defp label_from_value(atom) when is_atom(atom), do: atom
```
One more:

```suggestion
  defp label_from_value(ref) when is_reference(ref), do: inspect(ref)
```

josevalim commented:
@jonatanklosko the current implementation discards the result of the function given to trace and I don't like that. We could instead return `{res, kino_markdown}`, but then the result of course won't be printed. So here is my suggestion:

1. Have this function return both `{res, kino_markdown}`
2. Also add `Kino.Process.render_trace`, that calls `trace/2`, renders the trace behind the scenes, and returns only `res`

WDYT?

josevalim reviewed (commented):
Amazing @akoutmos! I have added only some minor final comments!

> Do you want me to remove all receive events and just have send events? 

Yes, I think we can remove them all. :)

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,212 @@ defmodule Kino.Process do
+    seq_trace([pid], trace_function)
+  end
+
+  def seq_trace(trace_pids, trace_function) do
```
```suggestion
  def seq_trace(trace_pids, trace_function) when is_list(trace_pids) do
```

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,212 @@ defmodule Kino.Process do
+
+    # Generate the mermaid formatted list of message events
+    messages =
+      trace_events
+      |> Enum.map_join("\n", fn {type, _timestamp, from, to, message} ->
```
It is a style thing but can you please remove the "single pipe" pipelines? More than one `|>` is fine tho. :)

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,212 @@ defmodule Kino.Process do
+    participants =
+      participants_lookup
+      |> Enum.map_join("\n", fn {pid, idx} ->
+        case Process.info(pid, :registered_name) do
+          nil ->
+            "participant #{idx} AS #35;PID#{:erlang.pid_to_list(pid)};"
+
+          {:registered_name, []} ->
+            "participant #{idx} AS #35;PID#{:erlang.pid_to_list(pid)};"
+
+          {:registered_name, name} ->
+            "participant #{idx} AS #{inspect(name)};"
+        end
```
```suggestion
        case Process.info(pid, :registered_name) do
          {:registered_name, name} when is_atom(name) ->
            "participant #{idx} AS #{inspect(name)};"

          _ ->
            "participant #{idx} AS #35;PID#{:erlang.pid_to_list(pid)};"
        end
```

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,212 @@ defmodule Kino.Process do
   end
 
+  @doc """
+  Generate a sequence diagram of process messages.
```
Now that you added `:all` and pids, it is important to clarify that the trace starts from the the current process.

```suggestion
  Generate a sequence diagram of process messages starting from `self()`.
```

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,212 @@ defmodule Kino.Process do
+        Agent.stop(agent_pid)
+      end)
+  """
+  @spec seq_trace(trace_target(), (() -> any())) :: Markdown.t()
```
```suggestion
  @spec seq_trace(trace_target(), (() -> any())) :: Markdown.t()
  def seq_trace(trace_target \\ :all, trace_function)
```

?

akoutmos replied:
The lazy programmer in me likes that :D

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,210 @@ defmodule Kino.Process do
+
+    # Generate the Mermaid formatted list of participants
+    participants =
+      participants_lookup
+      |> Enum.map_join("\n", fn {pid, idx} ->
```
```suggestion
      Enum.map_join(participants_lookup, "\n", fn {pid, idx} ->
```

josevalim commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,210 @@ defmodule Kino.Process do
+
+    # Get all the participating actors in the trace along with their sequence diagram IDs
+    {participants_lookup, _idx} =
+      trace_events
+      |> Enum.reduce({%{}, 0}, fn {_type, _timestamp, from, to, _message}, acc ->
```
```suggestion
      Enum.reduce(trace_events, {%{}, 0}, fn {_type, _timestamp, from, to, _message}, acc ->
```

akoutmos commented:
Thanks for the review @josevalim! Really appreciate it. I reworked the exception handling a little so that the call to the tracer GenServer was not part of the trace capture. I also need to rework that last reduce to handle the scenario when a process is started prior to the trace, but terminated during the trace. Mermaid throws an error if you attempt to deactivate an actor without it being activated prior.

After that I think it is good to go :). Thanks!

josevalim commented:
You can also do nested trys. The nesting is uglier but less code duplication and perhaps easier to understand?

```elixir
try do
  try do
    fun.()
  after
    stop_trace
  end

  get_trace_events
after
  GenServer.stop
end
```

akoutmos commented:
> You can also do nested trys. The nesting is uglier but less code duplication and perhaps easier to understand?
> 
> ```elixir
> try do
>   try do
>     fun.()
>   after
>     stop_trace
>   end
> 
>   get_trace_events
> after
>   GenServer.stop
> end
> ```

It has been done :)

akoutmos commented:
I fixed that last issue that was on my TODO list and switched the PR from draft to ready. Let me know if there are any other issues I should address.

I think the only open question at the moment is whether the result is kept from the trace function?

josevalim commented:
I will let you know soon. I also just realized that we could name the current process as `self()` in the diagram. WDYT?

akoutmos commented:
> I will let you know soon. I also just realized that we could name the current process as `self()` in the diagram. WDYT?

Sounds good :). I updated how the participating PIDs are generated and the calling PID now show up as `self()`. I think that makes it nice and clear for users following the trace.

josevalim commented:
Hi @akoutmos, let's please change this one to return `{res, kino}`. Then we will add render_app_tree, render_sup_tree and render_seq_trace that render as a side-effect. The first two will return `Kino.nothing`, the last one will return the result of the function. WDYT? Can you please push those final changes? :)

jonatanklosko commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,229 @@ defmodule Kino.Process do
+
+      Kino.Process.seq_trace(agent_pid, fn ->
+        1..2
+        |> Task.async_stream(fn value ->
+          Agent.get(agent_pid, fn value -> value end)
+          100 * value
+        end,
+        max_concurrency: 3)
```
```suggestion
        |> Task.async_stream(
          fn value ->
            Agent.get(agent_pid, fn value -> value end)
            100 * value
          end,
          max_concurrency: 3
        )
```

jonatanklosko commented on lib/kino/process.ex:
```diff
@@ -137,6 +139,229 @@ defmodule Kino.Process do
+        Process.monitor(agent_pid)
+
+        1..2
+        |> Task.async_stream(fn value ->
+          Agent.get(agent_pid, fn value -> value end)
+          100 * value
+        end,
+        max_concurrency: 3)
```
```suggestion
        |> Task.async_stream(
          fn value ->
            Agent.get(agent_pid, fn value -> value end)
            100 * value
          end,
          max_concurrency: 3
        )
```

jonatanklosko commented on lib/kino/process/tracer.ex:
```diff
@@ -0,0 +1,38 @@
+
+  use GenServer
+
+  def start_link(_) do
```
Since we start it by hand, we don't need an artificial argument :)

```suggestion
  def start_link() do
```

jonatanklosko commented on lib/kino/process/tracer.ex:
```diff
@@ -0,0 +1,38 @@
+
+  @impl true
+  def handle_info({:seq_trace, _, {:send, _, from, to, message}, timestamp}, trace_events) do
+    new_event = {:send, timestamp, from, to, message}
```
If we make it a map, it will likely make the matching cleaner later on, wdyt?

jonatanklosko commented on lib/kino/process/tracer.ex:
```diff
@@ -0,0 +1,38 @@
+    GenServer.start_link(__MODULE__, nil)
+  end
+
+  # ---- Client API ----
+
+  @doc false
+  def get_trace_events(tracer) do
+    GenServer.call(tracer, :get_trace_events)
+  end
+
+  # ---- Callbacks ----
```
```suggestion
  def get_trace_events(tracer) do
    GenServer.call(tracer, :get_trace_events)
  end
```

jonatanklosko reviewed (commented):
@akoutmos fantastic!!

akoutmos commented:
> @akoutmos fantastic!!

Thanks @jonatanklosko!

akoutmos commented:
> Hi @akoutmos, let's please change this one to return `{res, kino}`. Then we will add render_app_tree, render_sup_tree and render_seq_trace that render as a side-effect. The first two will return `Kino.nothing`, the last one will return the result of the function. WDYT? Can you please push those final changes? :)

That sounds good to me :). I think I implemented it as you described in [9a5b965](https://github.com/livebook-dev/kino/pull/165/commits/9a5b965cef23b8b89a939945a1cf42b4e4523ede). Let me know what you think!

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

akoutmos commented:
Thanks for the review @josevalim and @jonatanklosko! Happy to see this merged in :)
